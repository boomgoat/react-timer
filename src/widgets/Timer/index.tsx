import React from "react";
import { Button } from "../../components";
import { useTimer } from "../../hooks";
import { SpeedControl } from "../SpeedControls";

const Timer = () => {
  const {
    handleChange,
    handleClick,
    handleStart,
    setSpeed,
    speed,
    count,
    timerValue,
    isComplete,
    hasStartedTimer,
  } = useTimer();

  return (
    <>
      <div className="flex justify-center mb-4">
        <label
          data-testid="timer-input-label"
          htmlFor="timer-input"
          className=" font-bold text-2xl pt-3 mr-3"
        >
          Countdown:
        </label>
        <input
          data-testid="timer-input"
          type="number"
          id="timer-input"
          max={99}
          required
          className="text-black border-2 border-black px-3 w-40"
          value={timerValue}
          onChange={(e) => handleChange(e)}
        />
        <Button
          type="submit"
          dataTestId="start-button"
          classes="border-2 text-white bg-green-600 p-2 px-5 ml-4 m-1"
          clickEventHandler={() => handleStart()}
          label="Start"
        />
      </div>

      {count.minutes * 60 + count.seconds <= (parseInt(timerValue) * 60) / 2 ? (
        <h4 data-testid="halftime-marker" className="pt-12 italic">
          More than halfway there!
        </h4>
      ) : isComplete ? (
        <h4 data-testid="completed-marker" className="pt-12 italic">
          Time's up!
        </h4>
      ) : null}

      <div className="flex justify-center items-center">
        <div className=" w-96 mr-9">
          <h1
            data-testid="clock"
            className={
              "font-black text-9xl p-12 " +
              (count.minutes === 0 && count.seconds <= 10 && !isComplete
                ? "animate-flicker"
                : "text-black")
            }
          >
            {("0" + count.minutes).slice(-2) +
              ":" +
              ("0" + count.seconds).slice(-2)}
          </h1>
        </div>

        <div>
          <Button
            type="submit"
            dataTestId="pause-play-timer"
            clickEventHandler={() => handleClick()}
            classes="h-14 w-14 font-black text-xl pb-1 bg-white rounded-full border-black border-4"
            label={hasStartedTimer ? "| |" : <>&#9654;</>}
          />
        </div>
      </div>

      <SpeedControl speed={speed} setSpeed={setSpeed} />
    </>
  );
};

export default Timer;
