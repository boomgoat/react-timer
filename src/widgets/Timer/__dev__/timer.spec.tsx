import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import Timer from "..";

describe("Timer Tests", () => {
  test("Check if it renders correctly", () => {
    render(<Timer />);

    expect(screen.getByTestId("timer-input-label")).toBeInTheDocument();
    expect(screen.getByTestId("timer-input")).toBeInTheDocument();
    expect(screen.getByTestId("start-button")).toBeInTheDocument();
    expect(screen.queryByTestId("halftime-marker")).not.toBeInTheDocument();
    expect(screen.getByTestId("completed-marker")).toBeInTheDocument();
    expect(screen.getByTestId("clock")).toBeInTheDocument();
    expect(screen.getByTestId("pause-play-timer")).toBeInTheDocument();
  });

  test("Timer works as intended", () => {
    render(<Timer />);

    const timerInput = screen.getByTestId("timer-input");
    const startButton = screen.getByTestId("start-button");
    fireEvent.change(timerInput, { target: { value: "1" } });
    fireEvent.click(startButton);
    setTimeout(
      () => expect(screen.getByTestId("halfway-marker")).toBeInTheDocument(),
      1000 * 32
    );
    setTimeout(
      () => expect(screen.getByTestId("completed-marker")).toBeInTheDocument(),
      1000 * 60
    );
  });

  test("Timer input does not register characters other than numbers", () => {
    render(<Timer />);

    const timerInput: HTMLInputElement = screen.getByTestId("timer-input");
    fireEvent.change(timerInput, { target: { value: "as-e^" } });
    expect(timerInput.value).toBe("");
  });

  test("Timer input stops on pause", (done) => {
    render(<Timer />);

    const timerInput = screen.getByTestId("timer-input");
    const startButton = screen.getByTestId("start-button");
    const pauseButton = screen.getByTestId("pause-play-timer");
    fireEvent.change(timerInput, { target: { value: "1" } });
    fireEvent.click(startButton);

    const clock = screen.getByTestId("clock");
    setTimeout(() => {
      fireEvent.click(pauseButton);
      done();
      const pausedClock = screen.getByTestId("clock");
      console.log(pausedClock.textContent);
      expect(clock.textContent).toBe(pausedClock.textContent);
    }, 1000 * 4);
  });

  test("snapshot test", () => {
    const { asFragment } = render(<Timer />);
    expect(asFragment).toMatchSnapshot();
  });
});
