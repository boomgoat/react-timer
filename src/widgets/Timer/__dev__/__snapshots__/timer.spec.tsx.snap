// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer Tests snapshot test 1`] = `[Function]`;
